### PR TITLE
feat(indexer): Remaining tests for optimistic indexing

### LINKED ISSUE #6322: [indexer]: Add tests for optimistic indexing
We need to have tests for all the edge cases of optimistic indexing, to ensure that we store and return correct and consistent data.

Main testing scenarios are described here: https://github.com/iotaledger/iota/issues/6606#issuecomment-2934944278

### PR DESCRIPTION
**Pull request by tomxey**
# Description of change

Remaining test cases for optimistic indexing, as described in https://github.com/iotaledger/iota/issues/6606#issuecomment-2934944278

## Links to any relevant issues

fixes https://github.com/iotaledger/iota/issues/6322

## How the change has been tested

`cargo test --profile dev-nodebug --package iota-indexer --test rpc-tests --all-features -- write_api`

- [x] Basic tests (linting, compilation, formatting, unit/integration tests)
- [ ] Patch-specific tests (correctness, functionality coverage)

### Infrastructure QA (only required for crates that are maintained by @iotaledger/infrastructure)

- [ ] Synchronization of the indexer from genesis for a network including migration objects.
- [ ] Restart of indexer synchronization locally without resetting the database.
- [ ] Restart of indexer synchronization on a production-like database.
- [ ] Deployment of services using Docker.
- [ ] Verification of API backward compatibility.

### Release Notes

- [ ] Protocol:
- [ ] Nodes (Validators and Full nodes):
- [ ] Indexer:
- [ ] JSON-RPC:
- [ ] GraphQL:
- [ ] CLI:
- [ ] Rust SDK:
- [ ] REST API:

### CODE REVIEW

**Review by kodemartin — APPROVED** [review]
Looks good, but I would suggest revisiting the number of iterations for each optimsitic test case. 

The number currently ranges from 20 to 50 or even 150 for display updates. This increases the running time of the tests, and considering that these tests will run on every indexer change, it seems plausible to try to minimize and be consistent with the number of repetitions across tests.

File: crates/iota-indexer/tests/rpc-tests/write_api.rs [hunk 1]
```diff
@@ -36,6 +47,8 @@ use crate::{
 type TxBytes = Base64;
 type Signatures = Vec<Base64>;
 
+const NON_DETERMINISTIC_TESTS_REPETITIONS: usize = 20;
```
Comment (samuel-rufi) [inline crates/iota-indexer/tests/rpc-tests/write_api.rs:54]: nit: can you please add a small comment so that it's more clear what it reflects?

**Review by sergiupopescu199 — APPROVED** [review]
lgtm!

File: crates/iota-indexer/tests/rpc-tests/write_api.rs [hunk 2]
```diff
@@ -308,6 +321,155 @@ fn execute_transaction_block() {
...
+        runtime,
+        store,
+        cluster,
+        client,
+    } = ApiTestSetup::get_or_init();
+    runtime.block_on(async move {
+        indexer_wait_for_checkpoint(store, 1).await;
+        let (sender, sender_kp): (_, AccountKeyPair) = get_key_pair();
+
+        let gas = cluster
```
Comment (samuel-rufi) [inline crates/iota-indexer/tests/rpc-tests/write_api.rs]: nit: can you align this with `gas_ref` instead of `gas` as it's in the other tests

**Review by samuel-rufi — APPROVED** [review]
